notebooks/contrib-dev/Stochastic_Simu_Chem_Rxn.ipynb

```diff
@@ -0,0 +1,2654 @@
+        "\n",
+        "\\begin{equation}\n",
+        "\\frac{dy}{dx} = f(x, y)\n",
+        "\\end{equation}\n",
+        "\n",
+        "The Trapezoid Rule also operates through an iterative approach that discretizes the continuous ODE into discrete steps by using the following formula to update the solution at each step:\n",
+        "\n",
+        "\\begin{equation}\n",
+        "y_{n+1} = y_n + \\frac{h}{2} \\left[f(x_n, y_n) + f(x_{n+1}, y_{n+1})\\right]\n",
+        "\\end{equation}\n",
+        "\n",
+        "In this formula:\n",
+        "\n",
+        "$y_{n+1}$ is the new approximation of the solution at the next step.\n",
+        "\n",
+        "$y_n$ is the previous approximation of the solution at the current step.\n",
+        "\n",
+        "$h$ is the step size, representing the size of the interval over which the solution is approximated.\n",
+        "\n",
+        "$x_n$ is the current value of the independent variable at the nth step.\n",
+        "\n",
+        "$f(x_n, y_n)$ and $f(x_{n+1}, y_{n+1})$ represent the rate of change of y at the current point $(x_n, y_n)$ and the predicted point $(x_{n+1}, y_{n+1})$, respectively."
+      ],
+      "metadata": {
+        "id": "r1EYn2umKwWW"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "**Now, implement the found differential equations into the Crank-Nicolson (Trapezoid Rule) to find and plot the  numerically solution.**"
+      ],
+      "metadata": {
+        "id": "ZqnzYbpfnvHm"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "import numpy as np\n",
+        "import matplotlib.pyplot as plt\n",
+        "\n",
+        "# Define the parameters and initial conditions\n",
+        "k1 = 0.5\n",
+        "k2 = 1.0\n",
+        "xa0 = 1000.0  # Initial concentration of Ca\n",
+        "t_max = 16.0  # Maximum time\n",
+        "dt = 0.1  # Time step\n",
+        "\n",
+        "# Define the rate equations\n",
+        "RHS_Ca = lambda Ca, Cb, t: -k1 * Ca\n",
+        "RHS_Cb = lambda Ca, Cb, t: k1 * Ca - k2 * Cb\n",
+        "RHS_Cc = lambda Cb, t: k2 * Cb\n",
+        "\n",
+        "# Function to perform Crank-Nicolson method\n",
+        "def crank_nicolson(RHS_Ca, RHS_Cb, RHS_Cc, Ca0, Cb0, Cc0, dt, numsteps, LOUD=False):\n",
+        "    numsteps = int(numsteps)\n",
+        "    Ca = np.zeros(numsteps+1)\n",
+        "    Cb = np.zeros(numsteps+1)\n",
+        "    Cc = np.zeros(numsteps+1)\n",
+        "    t = np.arange(numsteps+1) * dt\n",
+        "    Ca[0] = Ca0\n",
+        "    Cb[0] = Cb0\n",
+        "    Cc[0] = Cc0\n",
+        "    for n in range(1, numsteps+1):\n",
+        "        if LOUD:\n",
+        "            print(\"\\nt =\", t[n])\n",
+        "\n",
+        "        # Crank-Nicolson Method\n",
+        "### BEGIN SOLUTION ###\n",
+        "        Ca_half = Ca[n-1] + 0.5 * dt * RHS_Ca(Ca[n-1], Cb[n-1], t[n])\n",
```

Comment:
This calculation is incorrect. Crank-Nicolson is an implicit method, meaning you must solve a (non)linear system of equations each iteration.

Reply:
https://ndcbe.github.io/data-and-computing/notebooks/07/Systems-of-Differential-Equations-and-Scipy.html#crank-nicolson